src/api/createRequest.js

```diff
@@ -0,0 +1,148 @@
+import axios from 'axios'
+
+// default headers are for json api
+const defaultHeaders = {
+  'Content-Type': 'application/vnd.api+json',
+  Accept: 'application/vnd.api+json',
+}
+
+/*
+ * @param {String} rootURL
+ * @param {String} apiNamespace
+ * @param {String} resource
+ * @param {Object} headers - these will default to json api headers
+ * @returns {Object} instance of request
+ */
+const createRequest = ({ rootURL, apiNamespace, resource, headers = defaultHeaders }) => {
+  const baseURL = `${rootURL}/${apiNamespace}`
+  const api = axios.create({ baseURL, headers })
+
+  /*
+   * @param {Any} value
+   * @returns boolean
+   */
+  const isObject = (value) => {
+    return value && typeof value === 'object' && value.constructor === Object
+  }
+
+  /*
+   * @param {Any} value
+   * @returns boolean
+   */
+  const isString = (value) => {
+    return (typeof value === 'string' || value instanceof String) && value.length > 0
+  }
+
+  /*
+   * @param {Object} attributes e.g. { parameter1: { key1: value1, key2: value2}, parameter2: { key1: value1, key2: value2 }, parameter3: value3}
+   * @return String e.g. 'parameter1[key1]=value1&parameter1[key2]=value2&parameter2[key1]=value1&parameter2[key2]=value2&parameter3=value3
+   */
+  const parametersToString = (attributes, parameter = undefined) => {
+    return Object.entries(attributes)
+      .flatMap(([key, value]) => {
+        // if the value is an object needs to be converted into a string
+        // we also pass the key as the parameter
+        if (isObject(value)) {
+          return parametersToString(value, key)
+        }
+
+        // if it is a string turn it into a parameter e.g. 'a.b.c' will become 'key=a.b.c'
+        // if parameter is defined it will be 'parameter[key]=value'
+        if (isString(value)) {
+          return parameter ? `${parameter}[${key}]=${value}` : `${key}=${value}`
+        }
```

Comment:
If we pass in non-string parameters here, such as a number/bool (which would be possible for a filter) then the parameter will be silently ignored. Should we throw if we match neither of these conditions to make it clear what's happened? (Part of me wonders if you just lose the string check entirely an let the string encoder do what it will.)

Reply:
I think I wanted to limit what could be passed and we could review later. Think I might just add a throw for now so at least it is not failing silently.